dependency-specification.rst

```diff
@@ -0,0 +1,376 @@
+implementations of the dependency specification which we can instead adopt.
+The existing implementations are battle proven and user friendly, so adopting
+them is arguably much better than approving an aspirational, unconsumed, format.
+
+Specification
+=============
+
+Examples
+--------
+
+All features of the language shown with a name based lookup::
+
+    requests \
+        [security] >= 2.8.1, == 2.8.* \
+        ; python_version < "2.7.10" \
+        # Fix HTTPS on older Python versions.
+
+A minimal URL based lookup::
+
+    pip @ https://github.com/pypa/pip/archive/1.3.1.zip#sha1=da9234ee9982d4bbb3c72346a6de940a148ea686
+
+Concepts
+--------
+
+A dependency specification always specifies a distribution name. It may
+include extras, which expand the dependencies of the named distribution to
+enable optional features. The version installed can be controlled using
+version limits, or giving the URL to a specific artifact to install. Finally
+the dependency can be made conditional using environment markers.
+
+Grammar
+-------
+
+We first cover the grammar briefly and then drill into the semantics of each
+section later.
+
+A distribution specification is written in ASCII text. We use ABNF [#abnf]_ to
+provide a precise grammar. Specifications may have comments starting with a
+'#' and running to the end of the line::
+
+    comment       = "#" *(WSP / VCHAR)
+
+Specifications may be spread across multiple lines if desired using
+continuations - a single backslash followed by a new line ('\\\\n')::
+
+    CSP           = 1*(WSP / ("\\" LF))
+
+Versions may be specified according to the PEP-440 [#pep440]_ rules. (Note:
+URI is defined in std-66 [#std66]_::
+
+    version-cmp   = "<" / "<=" / "!=" / "==" / ">=" / ">" / "~=" / "==="
+    version       = 1*( DIGIT / ALPHA / "-" / "_" / "." / "*" )
+    versionspec   = ["("] version-cmp version *(',' version-cmp version) [")"]
+    urlspec       = "@" URI
+
+Environment markers allow making a specification only take effect in some
+environments::
+
+    marker-op     = version-cmp / "in" / "not in"
+    python-str-c  = (WSP / ALPHA / DIGIT / "(" / ")" / "." / "{" / "}" /
+                    "-" / "_" / "*"
+    python-str    = "'" *(python-str-c / DQUOTE) "'"
+    python-str    =/ DQUOTE *(python-str-c / "'") DQUOTE
+    marker-var    = python-str / "python_version" / "python_full_version" /
+                    "os_name"" / "sys_platform" / "platform_release" /
+                    "platform_version" / "platform_machine" /
+                    "platform_python_implementation" / "implementation_name" /
+                    "implementation_version" / "platform_dist_name"
+                    "platform_dist_version" / "platform_dist_id"
+    marker-expr   = "(" marker ")" / (marker-var [marker-op marker-var])
+    marker        = marker-expr *( ("and" / "or") marker-expr)
+    name-marker   = ";" *CSP marker
+    url-marker    = ";" 1*CSP marker
```

Comment:
It guarantees that there will be at least one space after a url before a marker.